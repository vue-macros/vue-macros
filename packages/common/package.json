{
  "name": "@vue-macros/common",
  "version": "3.0.0-beta.4",
  "description": "common feature from Vue Macros.",
  "type": "module",
  "keywords": [
    "vue-macros",
    "macros",
    "vue",
    "sfc",
    "setup",
    "script-setup",
    "common"
  ],
  "license": "MIT",
  "homepage": "https://vue-macros.dev",
  "bugs": {
    "url": "https://github.com/vue-macros/vue-macros/issues"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vue-macros/vue-macros.git",
    "directory": "packages/common"
  },
  "author": "三咲智子 Kevin Deng <sxzz@sxzz.moe>",
  "funding": "https://github.com/sponsors/vue-macros",
  "files": [
    "dist"
  ],
  "main": "dist/index.js",
  "module": "dist/index.js",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "dev": "./src/index.ts",
      "default": "./dist/index.js"
    },
    "./*": "./*"
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": "./dist/index.js",
      "./*": "./*"
    },
    "tag": "next"
  },
  "scripts": {
    "build": "tsup",
    "dev": "DEV=true tsup"
  },
  "peerDependencies": {
    "vue": "^2.7.0 || ^3.2.25"
  },
  "peerDependenciesMeta": {
    "vue": {
      "optional": true
    }
  },
  "dependencies": {
    "@vue/compiler-sfc": "catalog:",
    "ast-kit": "^1.4.2",
    "local-pkg": "^1.1.1",
    "magic-string-ast": "^0.8.0",
    "unplugin-utils": "^0.2.4"
  },
  "devDependencies": {
    "@babel/parser": "catalog:",
    "@vitejs/plugin-vue": "catalog:"
  },
  "engines": {
    "node": ">=20.18.0"
  }
}
