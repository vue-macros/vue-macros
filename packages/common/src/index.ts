export * from 'magic-string-ast'
export * from 'ast-kit'
export {
  createFilter as createRollupFilter,
  normalizePath,
  type FilterPattern,
} from 'unplugin-utils'

export * from './ast'
export * from './constants'
export * from './dep'
export * from './options'
export * from './types'
export * from './unplugin'
export * from './vue'
