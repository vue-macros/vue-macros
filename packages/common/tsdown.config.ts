import { defineConfig } from 'tsdown/config'

export default defineConfig({
  entry: ['./src/index.ts'],
  external: ['node:module'],
})
