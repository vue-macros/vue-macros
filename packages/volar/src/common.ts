import path from 'node:path'
import process from 'node:process'
import {
  resolveOptions,
  type FeatureName,
  type OptionsResolved,
} from '@vue-macros/config'
import { replace, replaceAll, type Code } from 'ts-macro'
import type { SFCScriptBlock } from '@vue-macros/common'
import type { VueLanguagePlugin } from '@vue/language-core'

export const REGEX_DEFINE_COMPONENT: RegExp =
  /(?<=(?:__VLS_|\(await import\(\S+\)\)\.)defineComponent\(\{\n)/g

export function addProps(
  codes: Code[],
  decl: Code[],
  version: number,
): true | undefined {
  const codeString = codes.toString()
  if (!decl.length) return

  replace(
    codes,
    /(?<=type __VLS_PublicProps = )/,
    `{\n${decl.join(',\n')}\n} & `,
  )

  if (
    !codeString.includes(
      version >= 3.5
        ? '__typeProps: '
        : 'props: ,{} as __VLS_TypePropsToOption',
    )
  ) {
    replaceAll(
      codes,
      REGEX_DEFINE_COMPONENT,
      version >= 3.5
        ? '__typeProps: {} as __VLS_PublicProps,\n'
        : 'props: {} as __VLS_TypePropsToOption<__VLS_PublicProps>,\n',
    )

    if (version < 3.5 && !codeString.includes('type __VLS_NonUndefinedable')) {
      codes.push(
        `type __VLS_NonUndefinedable<T> = T extends undefined ? never : T;\n`,
        `type __VLS_TypePropsToOption<T> = { [K in keyof T]-?: {} extends Pick<T, K> ? { type: import('vue').PropType<__VLS_NonUndefinedable<T[K]>> } : { type: import('vue').PropType<T[K]>, required: true } };\n`,
      )
    }
  }

  return true
}

export function addEmits(
  codes: Code[],
  decl: Code[],
  version: number,
): true | undefined {
  const codeString = codes.toString()
  if (!decl.length) return

  const index = codes.findIndex((code) =>
    code.toString().startsWith('type __VLS_Emit = '),
  )
  const result = decl.join(',\n')
  const modelIndex = codes.findIndex((code) =>
    code.toString().includes('type __VLS_ModelEmit = '),
  )
  if (modelIndex !== -1) {
    codes.splice(modelIndex + 1, 0, ` ${result} , `)
  }
  if (index === -1) {
    const propsIndex = codes.findIndex((code) =>
      code.toString().includes('type __VLS_PublicProps = '),
    )
    codes.splice(
      propsIndex,
      0,
      ...(version < 3.5
        ? [
            `const __VLS_emit = (await import('vue')).defineEmits<{\n${result}\n}>();\n`,
            `type __VLS_Emit = typeof __VLS_emit;;\n`,
          ]
        : [`type __VLS_Emit = {\n${result}\n};\n`]),
    )
  } else if (modelIndex === -1) {
    codes.splice(index + 2, 0, ` & {\n${result}}\n`)
  }

  if (version >= 3.5) {
    if (!codeString.includes('__typeEmits: {} as ')) {
      replaceAll(
        codes,
        REGEX_DEFINE_COMPONENT,
        '__typeEmits: {} as __VLS_Emit,\n',
      )
    }
  } else if (!codeString.includes('{} as __VLS_NormalizeEmits<typeof __VLS')) {
    replaceAll(
      codes,
      REGEX_DEFINE_COMPONENT,
      'emits: {} as __VLS_NormalizeEmits<__VLS_Emit>,\n',
    )
  }

  return true
}

export function addCode(codes: Code[], ...args: Code[]): void {
  const index = codes.findIndex((code) =>
    code.includes('__VLS_setup = (async () => {'),
  )
  codes.splice(index !== -1 ? index + 1 : codes.length, 0, ...args)
}

export type VueMacrosPlugin<K extends FeatureName> = (
  ctx: PluginContext,
  options?: OptionsResolved[K],
) => ReturnType<VueLanguagePlugin>

export type PluginContext = Parameters<VueLanguagePlugin>[0]

const resolvedOptions: Map<string, OptionsResolved> = new Map()

export function getVolarOptions<K extends keyof OptionsResolved>(
  context: PluginContext,
  key: K,
): OptionsResolved[K] {
  const configPath = context.compilerOptions.configFilePath
  const root =
    typeof configPath === 'string' ? path.dirname(configPath) : process.cwd()
  let resolved: OptionsResolved | undefined
  if (!resolvedOptions.has(root)) {
    resolved = resolveOptions.sync(context.vueCompilerOptions.vueMacros, root)
    resolvedOptions.set(root, resolved)
  }

  return (resolved || resolvedOptions.get(root)!)[key]
}

export function patchSFC(block: SFCScriptBlock | null, offset: number): void {
  if (block) {
    block.loc.start.column -= offset
    block.loc.start.offset -= offset
    block.loc.end.offset -= offset
    if (block.loc.end.line === block.loc.start.line) {
      block.loc.end.column -= offset
    }
  }
}
