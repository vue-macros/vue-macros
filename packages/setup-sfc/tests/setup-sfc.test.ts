import { resolve } from 'node:path'
import { REGEX_SETUP_SFC } from '@vue-macros/common'
import {
  rollupBuild,
  RollupVue,
  RollupVueJsx,
  testFixtures,
  UnpluginOxc,
} from '@vue-macros/test-utils'
import { describe, expect, test } from 'vitest'
import VueSetupSFC from '../src/rollup'

describe('setup-component', () => {
  test('isSetupSFC', () => {
    expect(REGEX_SETUP_SFC.test('foo.setup.ts')).toBe(true)
    expect(REGEX_SETUP_SFC.test('foo.setup.tsx')).toBe(true)
    expect(REGEX_SETUP_SFC.test('foo.setup.jsx')).toBe(true)
    expect(REGEX_SETUP_SFC.test('foo.setup.js')).toBe(true)
    expect(REGEX_SETUP_SFC.test('foo.setup.mjs')).toBe(true)
    expect(REGEX_SETUP_SFC.test('foo.setup.cts')).toBe(true)

    expect(REGEX_SETUP_SFC.test('foo.ts')).toBe(false)
  })

  describe('fixtures', async () => {
    await testFixtures(
      'tests/fixtures/*.{vue,[jt]s?(x)}',
      (args, id) =>
        rollupBuild(id, [
          VueSetupSFC(),
          RollupVue({
            include: [REGEX_SETUP_SFC],
          }),
          RollupVueJsx(),
          UnpluginOxc.rollup(),
        ]),
      {
        cwd: resolve(__dirname, '..'),
        promise: true,
      },
    )
  })
})
