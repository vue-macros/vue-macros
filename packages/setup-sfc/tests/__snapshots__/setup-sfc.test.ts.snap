// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setup-component > fixtures > tests/fixtures/basic.setup.tsx 1`] = `
"// basic.setup.js
import { defineComponent, createVNode } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	__name: "basic.setup",
	props: { foo: {
		type: String,
		required: true
	} },
	setup(__props, { expose: __expose }) {
		__expose();
		defineRender(() => {
			// @ts-ignore
			return createVNode("div", null, null);
		});
		const __returned__ = {};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var basic_setup = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { basic_setup as default };
"
`;

exports[`setup-component > fixtures > tests/fixtures/with-define-options.setup.tsx 1`] = `
"// with-define-options.setup.js
import { defineComponent, createVNode } from 'vue';
import _export_sfc from '[NULL]/plugin-vue/export-helper';

var _sfc_main = /* @__PURE__ */ defineComponent({
	...{ name: "App" },
	__name: "with-define-options.setup",
	setup(__props, { expose: __expose }) {
		__expose();
		defineRender(() => {
			return createVNode("div", null, null);
		});
		const __returned__ = {};
		Object.defineProperty(__returned__, "__isScriptSetup", {
			enumerable: false,
			value: true
		});
		return __returned__;
	}
});

var withDefineOptions_setup = /* @__PURE__ */ _export_sfc(_sfc_main, [__FILE__]);

export { withDefineOptions_setup as default };
"
`;
